### PR TITLE
Writing tests - Draft

### PR DESCRIPTION
**Pull request by jiwalker-usgs**

### CODE REVIEW

File: content/r-package-dev/Writing_Tests.Rmd [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+    Sys.sleep(60) # calculating the answer to life, the universe, and everything
+    return(42)
+  }
+  unit_to_test <- function() {
+    answer <- really_long_running_function()
+    return(answer/6)
+  }
+  test_that("answer is seven", {
+    print(system.time({
+      with_mock(
```
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: wow, i totally need this and didn't know it was there. thanks, @jiwalker-usgs !

File: content/r-package-dev/Writing_Tests.Rmd [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+test_that("pH edge cases return true", {
+  expect_true(is.valid.pH(0))
+  expect_true(is.valid.pH(14))
+})
+```
+
+Through a bit of magic we can look at the results of these tests.  Normally this will happen by running "Build > Test Package" through RStudio (Ctrl+Shift+T shortcut).
+
+```{r test-output, include=TRUE, warning=FALSE, message=FALSE}
+  reporter$get_results() # only works for ListReporter
```
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: i like that the other `reporter` blocks are hidden and that you've noted the Ctrl+Shift+T option in the text. This particular code line shows up in the md. Can you hide it and just show the output with `echo=FALSE`?

**Comment by lindsaycarr** [comment]
:+1:  #155

Looks like it will work on dev, so merging. People will review & comment based on dev site.

File: content/r-package-dev/Writing_Tests.Rmd [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+**Data testing:** Data is often treated differently than code, but it can be tested in quite similar ways.  Assumptions about data types, valid ranges, and other aspects of the data should be explicitly tested for to ensure validity.
+
+Applying these techniques where applicable will develop a solid suite of tests that will instill confidence in those developing on or against your package.  It is helpful to gather code coverage metrics to quantify the level of testing that exists.  The R package `covr` is available to generate these metrics and push them to a location to share the results.
+
+## How do I write tests?
+
+We are going to build on what was covered in the documentation section by writing some tests.  Here is a reminder of the pH validity function we created earlier.
```
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: provide a link to the lesson where this function was first created?

File: content/r-package-dev/Writing_Tests.Rmd [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+In developing a package, you will be adding code to perform certain functions.  In order to prove that this code does what it claims to, write a test that runs the function with different inputs and verify the output (or side effects).  Test edge cases and error conditions to make sure they are handled appropriately.  Creating tests covering expected behavior and edge cases as part of a testing framework in your package ensures the following:
+
+**Document expected behavior:** Tests should clearly indicate what, given certain inputs, is expected to result from a given function.  This lays out the requirements for a piece of code and shows that it performs as specified.
+
+**Prevent regressions:** Everything is working right now, but in a month or so a new feature might be requested.  When adding this functionality there is a chance that interactions with other code might break an existing piece of functionality.  Having good tests in place will catch this quickly.
```
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: yep, i'm persuaded! :+1:

File: content/r-package-dev/Writing_Tests.Rmd [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+  pH >= 0 & pH <= 14
+}
+```
+
+```{r reporter-setup, include=FALSE, warning=FALSE, message=FALSE}
+  reporter <- testthat:::find_reporter("list")
+  old <- testthat:::set_reporter(reporter)
+```
+
+The `testthat` package provides the functions we are going to use for testing.  Files performing the testing should be placed in the `tests/testthat` directory and a `testhat.R` script should invoke `test_check()` on your package.
```
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: say where `testthat.R` needs to live

File: content/r-package-dev/Writing_Tests.Rmd [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+## Lesson Objectives
+
+1. Recognize the purpose and value of writing tests
+2. Understand different types of testing and when to apply them
+3. Learn some test writing practices and how often to run them
+
+
+## Why should I test my code?
+
+Formal testing of code is like showing your work on a hard math problem.  You may come to the correct solution, but unless you show exactly how you got to that solution you may not get full credit.  The same is true in software. Adding functionality to a package should be accompanied by code that exercises said functionality and asserts its validity.  Tests are also meant to signal to future developers what the requirements of particular functionality are and to warn when those are not being met.
```
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: hmm...almost. I like the presence of an analogy, but the math problem doesn't quite do it for me. tests aren't exactly the same as showing your work (the package code does that better than the tests). and this analogy successfully evokes the idea that tests are the grunt work part of development, but not sure we want that idea front and center...

how about "Trust, but verify"...or tests are like seatbelts (don't always seem essential, but protect you from big injuries at key times)...or something like that?

File: content/r-package-dev/Writing_Tests.Rmd (hunk 5 above)
Comment (aappling-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: also consider saying something about helper files that provide shared code across tests? i use them a lot, but not sure whether they'd be a key detail or a distraction for a new student/user of testthat...

File: content/r-package-dev/Writing_Tests.Rmd (hunk 1 above)
Comment (jiwalker-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: Yeah, I came across this recently and wanted to add it in.

File: content/r-package-dev/Writing_Tests.Rmd (hunk 2 above)
Comment (jiwalker-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: Good thought, I was trying to figure out how to minimize the "magic"

File: content/r-package-dev/Writing_Tests.Rmd (hunk 3 above)
Comment (jiwalker-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: yeah, we will want that link.

File: content/r-package-dev/Writing_Tests.Rmd (hunk 6 above)
Comment (jiwalker-usgs) [inline content/r-package-dev/Writing_Tests.Rmd]: I like the Reagan quote :)  I was trying to get to the Dykstra view of software, where it is more like math, but I see that the analogy doesn't quite connect.
